vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go

```diff
@@ -0,0 +1,275 @@
+	// stored at bucketForDay[(lastDayIndex-1)%retentionDays] etc.
+	lastDayIndex int
+	// Retention days.
+	retentionDays int
+}
+
+func (h *binaryDecayingHistogram) dayIndex(ts time.Time) int {
+	return int(ts.Unix() / (60 * 60 * 24))
+}
+
+func (h *binaryDecayingHistogram) addSampleToBucket(bucket uint16, dayIndex int) {
+	if h.lastDayIndex > 0 {
+		// Ignore samples too far in the past
+		if dayIndex <= h.lastDayIndex-h.retentionDays {
+			return
+		}
+		// If dayIndex not in the future, we will set bucket to max of current and new value
+		if dayIndex <= h.lastDayIndex && h.bucketForDay[dayIndex%h.retentionDays] > bucket {
+			bucket = h.bucketForDay[dayIndex%h.retentionDays]
+		}
+		// Move forward if necessary
+		if dayIndex > h.lastDayIndex {
+			i := h.lastDayIndex + 1
+			for i < dayIndex {
+				h.bucketForDay[i%h.retentionDays] = 0
+				// Break if we cleared out all days
+				if i%h.retentionDays == h.lastDayIndex%h.retentionDays {
+					break
+				}
+				i++
+			}
+		}
+	}
+	h.bucketForDay[dayIndex%h.retentionDays] = bucket
+	if dayIndex > h.lastDayIndex {
+		h.lastDayIndex = dayIndex
+	}
+}
+
+func (h *binaryDecayingHistogram) addSample(value float64, dayIndex int) {
+	bucket := h.options.FindBucket(value) + 1
+	h.addSampleToBucket(uint16(bucket), dayIndex)
+}
+
+func (h *binaryDecayingHistogram) AddSample(value float64, weight float64, time time.Time) {
+	dayIndex := h.dayIndex(time)
+	h.addSample(value, dayIndex)
+}
+
+func (h *binaryDecayingHistogram) SubtractSample(value float64, weight float64, time time.Time) {
+	panic("SubtractSample function is not implemented for binary decaying histogram")
+}
+
+func (h *binaryDecayingHistogram) Merge(other Histogram) {
+	o := other.(*binaryDecayingHistogram)
+	if h.options != o.options {
+		panic("can't merge histograms with different options")
+	}
+	if h.retentionDays != o.retentionDays {
+		panic("can't merge histograms with different retention days")
+	}
+	for day := o.lastDayIndex; day > o.lastDayIndex-o.retentionDays && day >= 0; day-- {
+		bucket := o.bucketForDay[day%o.retentionDays]
+		if day <= h.lastDayIndex && day > h.lastDayIndex-h.retentionDays {
+			if h.bucketForDay[day%h.retentionDays] > bucket {
+				bucket = h.bucketForDay[day%h.retentionDays]
+			}
+		}
+		if bucket > 0 {
+			h.addSampleToBucket(bucket, day)
+		}
+	}
+}
+
+func (h *binaryDecayingHistogram) Percentile(percentile float64) float64 {
+	if percentile != 1.0 {
+		panic("Only Percentile(1.0) function is implemented for binary decaying histogram, got: " + fmt.Sprintf("%f", percentile))
+	}
+	maxBucket := uint16(0)
+	for day := 0; day < h.retentionDays; day++ {
+		if h.bucketForDay[day] > maxBucket {
+			maxBucket = h.bucketForDay[day]
+		}
+	}
+	value := 0.0
+	if maxBucket > 0 {
+		value = h.valueForBucket(int(maxBucket) - 1)
+	}
+	return value
+}
+
+func (h *binaryDecayingHistogram) IsEmpty() bool {
```

Comment:
Out of curiosity, is this equivalent to checking if `lastDayIndex` is 0?

Reply:
I think you are right, they are equivalent, because we only change `lastDayIndex` when we add sample.